### DISCUSSION TITLE
How to extract multiple video streams at once?

### DISCUSSION THREAD

**Discussion by cansik** [python-ffmpegio/python-ffmpegio#17] [Q&A]
I have an Azure Kinect MKV file which I would like to extract frames from. The file contains a RGB video stream (`rgb24`) and two `gray16le` streams for the IR and Depth values. In the documentation I have read that with `vv` it would be possible to read all video streams, but it seems that only the RGB stream is extracted. Or I use the `map=0:1` method to extract a specific stream. 

```python
def read_streams(input_path: Path):
    with ffmpegio.open("test.mkv", 'vv', blocksize=1) as fin:
        for frames in fin:
            print(frames.keys())
```

How is it possible to read all streams at once?

Attached: [test.mkv](https://github.com/microsoft/Azure-Kinect-Samples/blob/master/body-tracking-samples/sample_recordings/test.mkv)

**Comment by cansik** [comment]
Ok, it seems that `map` can be applied multiple times:

```python
with ffmpegio.open(str(input_path), "rvv", map=["0:0", "0:1", "0:2"], blocksize=1) as fin:
    # ... returns dictionary with `v:0`, `v:1` and `v:2`
```

But now the resulting video streams are not encoded the correct way. Is it possible to specify different encodings per stream? I think it should be possible using the `split` filter, but I can't get it right atm. Would be great if someone would have a hint for me.

**Reply by cansik** [reply]
Another try was to later re-interpret the data with numpy, but this leads to an image which seems not to contain the right width and height:

```python
 v2: np.ndarray = frames["v:2"]
 data = v2.reshape(-1).view(np.uint16).reshape([-1, 576, 640, 1])  # 640 doesn't work, it has to be 960
```

**Reply by tikuma-lsuhsc** [reply]
@cansik - 

First of all, here is my test code

```python
import ffmpegio as ff

input_path = "sandbox/test_cansik.mkv"
streams = ["v:0", "v:1", "v:2"]

sinfo = ff.probe.video_streams_basic(input_path)

# print(ff.media.read(input_path, t=1, map=streams))
with ff.open(str(input_path), "rvv", map=streams, blocksize=1) as fin:
    F = fin.read()

    print(list(F.keys()))
    for i, s in enumerate(streams):
        print(s, F[s].shape, (1, sinfo[i]["width"], sinfo[i]["height"], 3))
        print(s, F[s].dtype, sinfo[i]["pix_fmt"])
```

this prints

```
['v:0', 'v:1', 'v:2']
v:0 (1, 1080, 1920, 3) (1, 1920, 1080, 3)
v:0 uint8 yuvj422p
v:1 (1, 576, 640, 3) (1, 640, 576, 3)
v:1 uint8 gray16be
v:2 (1, 576, 640, 3) (1, 640, 576, 3)
v:2 uint8 gray16be
```

and your issue is that the output frame `dtype` is not correct for the `v:1` and `v:2`. Am I correct?

There is a way but with a caveat that there is a bug that needs to be fixed first (or rather current incompatibility with FFmpeg). The nature of this generic media reader beast is that the output video format defaults to RGB24. In other words, it appends `-pix_fmt rgb24` when `ffmpegio` calls FFmpeg. 

In FFmpeg, what you'd do to override this is to specify per output stream: `-pix_fmt:1 gray16le` and `-pix_fmt:2 gray16le` (using the little endian format which is native to  x86's). This translates to `ffmpegio` as 

```python
pix_fmts = {"pix_fmt:v:0": "rgb24", "pix_fmt:v:1": "gray16le", "pix_fmt:v:2": "gray16le"}
with ff.open(str(input_path), "rvv", map=streams, blocksize=1, **pix_fmts) as fin:
    ...
```

However, the said bug currently appends `-pix_fmt rgb24` to the end of the FFmpeg command line, overriding per-stream specification because of how `ffmpegio` orders its FFmpeg option arguments... Anyway, a long story short, I need to fix this bug for you.

Meanwhile, if you'd like to hack your way through, look into `Lib/site-packages/ffmpegio_core/configure.py` and find Line 623-4 (in `finalize_media_read_opts()`):

```
    if "pix_fmt" not in options:
        options["pix_fmt"] = "rgb24"
```

You can comment out these 2 lines while providing the above `pix_fmts` keyword arguments, you should be able to get it to work. Meanwhile, I'll try to push the fix out over the weekend.

**Reply by cansik** [reply]
Oh wow, thanks for your very detailed analysis of the problem, and yes, that is exactly the issue I encounter. Looking forward to a fix 🙌🏻 

If you don't find the time, maybe I could try working on a PR. But I think it would take way longer, since I am not into the codebase.

**Reply by cansik** [reply]
@tikuma-lsuhsc As a workaround, I found out that it is possible to provide a default `pix_fmt`, as well as the channel specific ones, which seems to work as well (instead of commenting out the check):

```python
with ffmpegio.open(str(input_path), "rvv", map=streams, blocksize=1, pix_fmt="rgb24", **pix_fmts) as fin:
    ...
```

**Reply by tikuma-lsuhsc** [reply]
Nice find! And that's interesting. I guess `dict` is aware of the orders keys are assigned. Good news. That'll make our life a bit easier :)

Just FYI, I created this generic multi-stream reader just out of curiosity without any specific use case for myself (for now). If you encounter any issues, feel free to post those issues on Issues board and I'll try to address'em as quickly as possible.

**Comment by tikuma-lsuhsc** [comment]
@cansik - FYI, just released v0.9.0, which includes the fix. Give it a try!

**Reply by cansik** [reply]
Great, I will check it out tomorrow! I am currently working on writing an open recording reader for the Azure Kinect SDK, as the SDK is currently only supported on Windows & Linux. However, sometimes it would be great to be able to read recordings on MacOS. And the multi-stream reader is perfect for that :)

**Reply by tikuma-lsuhsc** [reply]
Cool. MacOS is the least tested environment. So, if you encounter any funniness, please report. Thanks!

Good luck with your dev.

**Reply by cansik** [reply]
hey @tikuma-lsuhsc Thanks again for your help regarding reading the MKV data, it works like charm.

Now I face another issue, I would like to read the frames frame by frame. But it  seems that ffmpeg is decoding it in the background without me being able to process them quick enough:

```python
import time

import ffmpegio as ff

input_path = "test.mkv"
streams = ["v:0", "v:1", "v:2"]

i = 0
with ff.open(str(input_path), "rvv", map=streams, blocksize=1) as fin:
    while True:
        F = fin.read()
        print(f"{i} = {list(F.keys())}")
        time.sleep(0.1)

        i += 1
```

When I sleep  `0.01` seconds to simulate processing time,  the output written is the following (the clip has more than 38 frames). it seems that the iterator is becoming invalid as soon as ffmpeg has finished decoding the frames.

```python
...
35 = ['v:0', 'v:1', 'v:2']
36 = ['v:0', 'v:1', 'v:2']
37 = ['v:0', 'v:1', 'v:2']
38 = ['v:0', 'v:1', 'v:2']
Traceback (most recent call last):
  File "test.py", line 14, in <module>
    F = fin.read()
        ^^^^^^^^^^
  File ".venv/lib/python3.11/site-packages/ffmpegio/streams/AviStreams.py", line 220, in read
    return self._reader.read(n, ref_stream, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".venv/lib/python3.11/site-packages/ffmpegio/threading.py", line 608, in read
    raise TimeoutError("timed out waiting for the stream headers")
TimeoutError: timed out waiting for the stream headers
```

Is this a bug or a feature, and can this be prevented? I already thought the blocksize of `1` would prevent that behaviour. The received frames from the iterator are not skipping, so I think it is more a bug or something I don't understand yet.

Or second, is it possible to either loop the video decoding so the iterator won't be invalid? I tried with `sp_kwargs={"stream_loop", -1}` but this doesn't seem to work.

**Reply by tikuma-lsuhsc** [reply]
@cansik - 

> Is this a bug or a feature, and can this be prevented? I already thought the blocksize of 1 would prevent that behaviour. The received frames from the iterator are not skipping, so I think it is more a bug or something I don't understand yet.

It's 100% a bug. So, let me get a few things straight, making sure I understand your observation.

* The example loop works if you reduce the sleep time sufficiently short?
* "the blocksize of 1" Did this actually help? (I'd assume the opposite and go larger, but that may not resolve the issue.)

I may need to fast-track the planned transition from AVI based multi-stream reader to a named-pipe based system, which I think would be more stable. Nonetheless, I'll see if I can find time over the next few days to look at this closely.

> Or second, is it possible to either loop the video decoding so the iterator won't be invalid? I tried with sp_kwargs={"stream_loop", -1} but this doesn't seem to work.

I presume that you're wanting to put the video on an infinite loop by commanding FFmpeg `stream_loop=-1`. What you've tried is close but won't work because `stream_loop` is an input option. Hence, you need to use `stream_loop_in=-1` in `ffmpegio` so its cli command builder knows it's an input option. Give it a try.

**Reply by cansik** [reply]
Yes, `stream_loop_in=-1` seems to work as expected. This is already a good workaround for me, because this prevents the iterator to be invalid.

Regarding the issue, in the [example MKV](https://github.com/microsoft/Azure-Kinect-Samples/blob/master/body-tracking-samples/sample_recordings/test.mkv) I use for testing, there are `450` frames in the RGB stream (when using `readall()` on the `AviMediaReader` object). The loop always stops at `199`, which seems to me to be an arbitrary number after the stream is closed and the iterator is no longer valid. The sleep just makes it happen sooner.

I will read a bit more about the blocksize. `1` seems to be an odd number, but it was necessary to set it because for me it controls how many frame sets (n-frames per stream) are returned. Here for example `blocksize=2`:

<img width="402" alt="image" src="https://github.com/python-ffmpegio/python-ffmpegio/assets/5220162/3f57c906-302c-402e-b707-379bb437f1c5">

**Reply by tikuma-lsuhsc** [reply]
> The loop always stops at 199

If it's always 199, it's not arbitrary, no doubt. I'll look at your test vid tonight, but are the framerates different on the latter 2 streams? I see you only got 1 frame from them while getting 3 on the first. It could be a bug exposed by the heterogeneous framerates. (something I never tested)

> it was necessary to set it because for me controls how many frame sets (n-frames per stream)

OK. Your description above sounded like you changed to 1 to make the loop work.

The test vid will be a great help. I'll keep you posted.

**Reply by cansik** [reply]
The frame rate is 30/1 on all streams, but it is possible for the streams to be out of sync because they come from different sensors. There are `452` frames for the infrared and depth streams (the two monocular channels). Maybe it's just a really bad example, because usually it makes sense to synchronise them anyway before writing them into a recording file. If you need any further information, just leave a comment.

**Reply by tikuma-lsuhsc** [reply]
Oh, doh, I misread the array dimension. Yes, they all have 2 frames each... This should be enough information to debug the issue. Thanks for the report!

**Reply by tikuma-lsuhsc** [reply]
Just release v0.10.0 with the fix. It also has an enhancement which you may want to take advantage of:

You can now check the validity (not EOF) of the ffmpeg stream with its `__bool__` type caster:

```python
import time

import ffmpegio as ff

input_path = "test.mkv"
streams = ["v:0", "v:1", "v:2"]

i = 0
with ff.open(str(input_path), "rvv", map=streams, blocksize=1) as fin:
    while fin:
        F = fin.read()
        print(f"{i} = {list(F.keys())}")
        time.sleep(0.1)

        i += 1
```

Also, I fixed the bug in `__next__` so you can also do this: 


```python
with ff.open(str(input_path), "rvv", map=streams, blocksize=1) as fin:
    for F in fin:
        print(f"{i} = {list(F.keys())}")
        time.sleep(0.1)

        i += 1
```

and it'll terminate properly.

**Comment by tikuma-lsuhsc** [comment]
One other thing is the internal queue size. By default, the queue size is set to infinite so that the reader's thread can take every frame ffmpeg produces as fast as possible and shove it in the queue for the main thread to retrieve the frames. This however is bloats the memory footprint unnecessary when your processing speed is far slower. Though undocumented, you can set the queue size in the constructor for the multi-stream reader (and only multi-stream reader for now). For example,

```python
with ff.open(str(input_path), "rvv", map=streams, blocksize=1, queuesize=6) as fin:
    ...
```

This effectively gives you a double buffer (you're reading 3 frames per block from 3 streams) so the reader thread can stuff the queue one block ahead of your main process. You can make it larger as you wish. But I'd keep it at minimum double the number of frames that you are reading at a time.

**Comment by cansik** [comment]
@tikuma-lsuhsc Thanks again for your help, I finally managed to release the library: https://github.com/cansik/open-azure-kinect 

There are still a few things I wanted to ask you, mainly regarding searching and timecode retrieval.

1. Is it possible to `seek` within the stream by navigating to a specific position in the stream?
2. The Azure Kinect adds a timecode in usec to the frames. Is it possible to retrieve frame-based metadata?

**Reply by tikuma-lsuhsc** [reply]
> Is it possible to seek within the stream by navigating to a specific position in the stream?

No. You can only open a stream at a specific position. For that degree of control, you need to turn to `pyav` in which you're responsible to operate FFmpeg backend (thus having the full control over frame navigation).

If you want to stick with `ffmpegio` (or FFmpeg-as-a-subprocess approach) then your best bet is to add an abstraction layer, and reopen the `ffmpegio` stream in the background to operate a seek operation while the frontend stream object of yours maintains a uniform interface to outside. No, not the most elegant solution, but this is the limitation of this approach.

> The Azure Kinect adds a timecode in usec to the frames. Is it possible to retrieve frame-based metadata?

Theoretically, yes, but I need to experiment with it first to make sure. Do you have a *short* test video that I can use to tinker? If so please post here (or a link). Also, do you need it along the video stream or can you read ahead of time?

**Reply by tikuma-lsuhsc** [reply]
@cansik - I looked at your repo, and I have a couple comments

* license. Because this repo is GPL-v2, my understanding is that your work, which is a derivative (because it uses `ffmpegio`) also needs to be GPL-v2.
* Missing `opencv-python` in `requirements.txt` (though I'm always curious if you can reduce a package footprint by using FFmpeg filters rather than relying on OpenCV, but I digress)
* Use of `cv2.cvtColor(data[0], cv2.COLOR_BGR2RGB)`. This shouldn't be needed unless you want to flip red and blue. `Ffmpegio` by default reads all video frames as `'rgb24'`. Are you sure you need it?

**Reply by cansik** [reply]
@tikuma-lsuhsc Thanks, good point about the licence. I must have overlooked it as my template uses the MIT licence by default. I have updated it and will release a new version. Just out of curiosity, what is the reason for releasing ffmpegio under GPL-v2 and not LGPL or something less permissive?

The BGR2RGB conversion is needed, maybe it's because the Kinect stores the frames in bgr24. But I will have a look at it later, thanks for mentioning it.

A test video can be found here: https://github.com/microsoft/Azure-Kinect-Samples/blob/master/body-tracking-samples/sample_recordings/test.mkv

**Reply by tikuma-lsuhsc** [reply]
>The BGR2RGB conversion is needed, maybe it's because the Kinect stores the frames in bgr24.

I may know why that's happening (though I don't know how I overlooked it) and if so it is an bug in `ffmpegio`. Let me check on it.

**Reply by tikuma-lsuhsc** [reply]
Nope, it (RGB) is working as intended. Here is what the first frame looks

![test_cansik_Kinect](https://github.com/user-attachments/assets/bccbeb09-e616-4f67-84e8-93db6e02c86c)

This is plotted by

```python
plt.imshow(F['v:0'][0,...])
```
So `F['v:0'][0,:,:,0]` channel must be red.

**Reply by cansik** [reply]
I think matplotlib works with RGB and opencv by default uses BGR. I can't test it right now, but I think this could be the reason why I do the swap.

**Reply by tikuma-lsuhsc** [reply]
If that's the case, maybe `cv2.COLOR_RGB2BGR` is the right macro to use? Furthermore, you can pass the keyword argument `**{'pix_fmt:v:0': 'bgr24'}` to FFmpeg to avoid the need of the conversion.

**Comment by tikuma-lsuhsc** [comment]
>The Azure Kinect adds a timecode in usec to the frames. Is it possible to retrieve frame-based metadata?

I don't see any frame metadata embedded in any of the video streams in `test.mkv`... (FYI, you get the frame metadata by running `metadata` filter in `print` mode.)

But, I can postulate where to find this information. The main video stream is encoded in [Motion JPEG](https://en.wikipedia.org/wiki/Motion_JPEG) which is an intraframe-compression (de fact) "standard" so it stores a sequence of JPEG images with their capture time slapped on as their labels. As such, it supports non-uniform frame rates. Moreover, the time base of the stream is 1e6 (frame time is counted in 1-us interval, which corresponds to your description of the expected timecode resolution. 

So, the bottom line is use `ffprobe`'s `-show_frame` mode. `ffmpegio` has a wrapper for this that you can use:

```python
tlabel = 'best_effort_timestamp_time'
t = np.array(
    [
        f[tlabel]
        for f in ff.probe.frames(
            input_path, streams="v:0", entries=["best_effort_timestamp_time"]
        )
    ]
)
```

The safest field to get the time is `best_effort_timestamp_time` (though all the other time stamps should be the same for MJPEG). It does take a little while as it reads the entire video file.

If you want to see other fields, run

```python
ff.probe.frames(input_path, streams="v:0", intervals=1)
```

to get the info of the first frame. (`intervals` is a finicky argument btw: `int` to read N frames, `float` to read up to T seconds, and more complex options)
